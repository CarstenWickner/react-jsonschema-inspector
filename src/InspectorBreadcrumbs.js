import PropTypes from "prop-types";
import React, { PureComponent } from "react";
import classNames from "classnames";

import JsonSchemaPropType from "./JsonSchemaPropType";
import { getTypeOfArrayItems, isDefined } from "./utils";

class InspectorBreadcrumbs extends PureComponent {
    renderSingleColumnSelection({
        items, selectedItem, trailingSelection, onSelect
    }, index, columnData) {
        if (!selectedItem) {
            // no selection: nothing to show in the breadcrumbs
            return null;
        }
        const {
            prefix, separator, arrayItemAccessor, refTargets, preventNavigation
        } = this.props;
        // if this is not the last selection, we need to add a (dummy) array item selector for each nested array
        let suffix = "";
        if (!trailingSelection) {
            let itemSchema = getTypeOfArrayItems(items[selectedItem], refTargets);
            while (isDefined(itemSchema)) {
                suffix += arrayItemAccessor;
                itemSchema = getTypeOfArrayItems(itemSchema, refTargets);
            }
        }
        const className = classNames({
            "jsonschema-inspector-breadcrumbs-item": true,
            "has-nested-items": index < (columnData.length - 1)
        });
        return (
            <span
                key={index}
                className={className}
                onDoubleClick={preventNavigation ? undefined : event => onSelect(event, selectedItem)}
            >
                {(index === 0 ? prefix : separator) + selectedItem + suffix}
            </span>
        );
    }

    render() {
        const { columnData } = this.props;
        return (
            <div className="jsonschema-inspector-breadcrumbs">
                <span className="jsonschema-inspector-breadcrumbs-icon" />
                {columnData.map(this.renderSingleColumnSelection.bind(this))}
            </div>
        );
    }
}

InspectorBreadcrumbs.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(PropTypes.oneOfType([PropTypes.bool, JsonSchemaPropType])).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool,
        onSelect: PropTypes.func.isRequired
    })).isRequired,
    refTargets: PropTypes.objectOf(JsonSchemaPropType).isRequired,
    prefix: PropTypes.string,
    separator: PropTypes.string,
    arrayItemAccessor: PropTypes.string,
    preventNavigation: PropTypes.bool
};

InspectorBreadcrumbs.defaultProps = {
    prefix: "",
    separator: ".",
    arrayItemAccessor: "[0]",
    preventNavigation: false
};

export default InspectorBreadcrumbs;
