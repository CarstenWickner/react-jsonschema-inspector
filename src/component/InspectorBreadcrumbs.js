import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";

import { getColumnDataPropTypeShape } from "./renderDataUtils";

import createBreadcrumbBuilder from "../model/breadcrumbsUtils";

const InspectorBreadcrumbs = ({ columnData, breadcrumbsOptions }) => {
    const buildBreadcrumb = createBreadcrumbBuilder(breadcrumbsOptions);
    const { preventNavigation } = breadcrumbsOptions;
    return (
        <div className="jsonschema-inspector-breadcrumbs">
            <span className="jsonschema-inspector-breadcrumbs-icon" />
            {columnData.map((column, index) => {
                const breadcrumbText = buildBreadcrumb(column, index);
                if (!breadcrumbText) {
                    // omit empty breadcrumb
                    return null;
                }
                const className = classNames({
                    "jsonschema-inspector-breadcrumbs-item": true,
                    "has-nested-items": index < (columnData.length - 2)
                        || (index === (columnData.length - 2)
                            && (!columnData[columnData.length - 1].options || !columnData[columnData.length - 1].selectedItem))
                });
                const { selectedItem, onSelect } = column;
                return (
                    <span
                        // eslint-disable-next-line react/no-array-index-key
                        key={index}
                        className={className}
                        onDoubleClick={preventNavigation ? undefined : event => onSelect(event, selectedItem)}
                    >
                        {breadcrumbText}
                    </span>
                );
            })}
        </div>
    );
};

InspectorBreadcrumbs.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
    breadcrumbsOptions: PropTypes.shape({
        prefix: PropTypes.string,
        separator: PropTypes.string,
        skipSeparator: PropTypes.func,
        mutateName: PropTypes.func,
        preventNavigation: PropTypes.bool
    }).isRequired
};

export default InspectorBreadcrumbs;
