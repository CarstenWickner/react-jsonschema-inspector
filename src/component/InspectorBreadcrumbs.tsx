import * as PropTypes from "prop-types";
import * as React from "react";
import classNames from "classnames";

import { ColumnDataPropType } from "./renderDataUtils";
import { createBreadcrumbBuilder } from "../model/breadcrumbsUtils";
import { BreadcrumbsOptions, RenderColumn, RenderOptionsColumn } from "../types/Inspector";

interface InspectorBreadcrumbsProps {
    columnData: Array<RenderColumn>;
    breadcrumbsOptions: BreadcrumbsOptions;
}

export class InspectorBreadcrumbs extends React.Component<InspectorBreadcrumbsProps> {
    render(): React.ReactNode {
        const { columnData, breadcrumbsOptions } = this.props;
        const buildBreadcrumb = createBreadcrumbBuilder(breadcrumbsOptions);
        const { preventNavigation, renderItem, renderTrailingContent } = breadcrumbsOptions;
        return (
            <>
                <div className="jsonschema-inspector-breadcrumbs">
                    <span className="jsonschema-inspector-breadcrumbs-icon" />
                    {columnData.map((column, index) => {
                        const breadcrumbText = buildBreadcrumb(column, index);
                        if (!breadcrumbText) {
                            // omit empty breadcrumb
                            return null;
                        }
                        const hasNestedItems =
                            index < columnData.length - 2 ||
                            (index === columnData.length - 2 &&
                                (!(columnData[columnData.length - 1] as RenderOptionsColumn).options ||
                                    !columnData[columnData.length - 1].selectedItem));
                        if (renderItem) {
                            return renderItem({
                                breadcrumbText,
                                hasNestedItems,
                                column,
                                index
                            });
                        }
                        const { selectedItem, onSelect } = column;
                        return (
                            <span
                                key={index}
                                className={classNames({
                                    "jsonschema-inspector-breadcrumbs-item": true,
                                    "has-nested-items": hasNestedItems
                                })}
                                onDoubleClick={preventNavigation ? undefined : (event): void => onSelect(event, selectedItem)}
                            >
                                {breadcrumbText}
                            </span>
                        );
                    })}
                </div>
                {renderTrailingContent &&
                    renderTrailingContent({
                        breadcrumbTexts: columnData.map(buildBreadcrumb).filter((b) => !!b),
                        columnData
                    })}
            </>
        );
    }

    static propTypes = {
        columnData: ColumnDataPropType.isRequired,
        breadcrumbsOptions: PropTypes.shape({
            prefix: PropTypes.string,
            separator: PropTypes.string,
            skipSeparator: PropTypes.func,
            mutateName: PropTypes.func,
            preventNavigation: PropTypes.bool,
            renderItem: PropTypes.func,
            renderTrailingContent: PropTypes.func
        }).isRequired
    };
}
