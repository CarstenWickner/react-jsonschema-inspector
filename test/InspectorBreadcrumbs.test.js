import React from "react";
import { shallow } from "enzyme";
import InspectorBreadcrumbs from "../src/InspectorBreadcrumbs";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const schema = {
            properties: {
                "Item One": true,
                "Item Two": true
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: { "Schema One": schema },
                        selectedItem: "Schema One",
                        trailingSelection: true,
                        onSelect: () => { }
                    },
                    {
                        items: schema.properties,
                        selectedItem: null,
                        trailingSelection: false,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("without selection", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: { root: {} },
                        selectedItem: null,
                        trailingSelection: false,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-icon").exists()).toBe(true);
        expect(component.find(".jsonschema-inspector-breadcrumbs-item").exists()).toBe(false);
        expect(component.text()).toEqual("");
    });
    it("with prefix", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: { root: {} },
                        selectedItem: "root",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
                prefix="variableContext."
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(1);
        expect(component.text()).toEqual("variableContext.root");
    });
    it("with root array selection", () => {
        const itemSchema = {
            properties: {
                itemProperty: { title: "Property Title" }
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: {
                            root: { items: itemSchema }
                        },
                        selectedItem: "root",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(1);
        expect(component.text()).toEqual("root");
    });
    it("with array item selection", () => {
        const itemSchema = {
            properties: {
                itemProperty: { title: "Property Title" }
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: {
                            root: { items: itemSchema }
                        },
                        selectedItem: "root",
                        trailingSelection: false,
                        onSelect: () => { }
                    },
                    {
                        items: itemSchema.properties,
                        selectedItem: "itemProperty",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
        expect(component.text()).toEqual("root[0].itemProperty");
    });
    it("with custom arrayItemAccessor", () => {
        const itemSchema = {
            properties: {
                itemProperty: { title: "Property Title" }
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: {
                            root: { items: itemSchema }
                        },
                        selectedItem: "root",
                        trailingSelection: false,
                        onSelect: () => { }
                    },
                    {
                        items: itemSchema.properties,
                        selectedItem: "itemProperty",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
                arrayItemAccessor=".get(0)"
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
        expect(component.text()).toEqual("root.get(0).itemProperty");
    });
    it("with nested array item selection", () => {
        const itemSchema = {
            properties: {
                itemProperty: { title: "Property Title" }
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: {
                            root: {
                                items: {
                                    items: {
                                        items: itemSchema
                                    }
                                }
                            }
                        },
                        selectedItem: "root",
                        trailingSelection: false,
                        onSelect: () => { }
                    },
                    {
                        items: itemSchema.properties,
                        selectedItem: "itemProperty",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
        expect(component.text()).toEqual("root[0][0][0].itemProperty");
    });
    it("with custom separator", () => {
        const itemSchema = {
            properties: {
                itemProperty: { title: "Property Title" }
            }
        };
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: {
                            root: {
                                properties: {
                                    item: itemSchema
                                }
                            }
                        },
                        selectedItem: "root",
                        trailingSelection: false,
                        onSelect: () => { }
                    },
                    {
                        items: {
                            item: itemSchema
                        },
                        selectedItem: "item",
                        trailingSelection: true,
                        onSelect: () => { }
                    },
                    {
                        items: itemSchema.properties,
                        selectedItem: null,
                        trailingSelection: false,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
                prefix="$this->"
                separator="->"
            />
        );
        expect(component.find(".jsonschema-inspector-breadcrumbs-item")).toHaveLength(2);
        expect(component.text()).toEqual("$this->root->item");
    });
    it("with navigation by default", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: { root: {} },
                        selectedItem: "root",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
            />
        );
        const selectedRootItem = component.find(".jsonschema-inspector-breadcrumbs-item");
        expect(selectedRootItem).toHaveLength(1);
        expect(selectedRootItem.prop("onDoubleClick")).toBeDefined();
    });
    it("when preventing navigation", () => {
        const component = shallow(
            <InspectorBreadcrumbs
                columnData={[
                    {
                        items: { root: {} },
                        selectedItem: "root",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                refTargets={{}}
                preventNavigation
            />
        );
        const selectedRootItem = component.find(".jsonschema-inspector-breadcrumbs-item");
        expect(selectedRootItem).toHaveLength(1);
        expect(selectedRootItem.prop("onDoubleClick")).not.toBeDefined();
    });
});
