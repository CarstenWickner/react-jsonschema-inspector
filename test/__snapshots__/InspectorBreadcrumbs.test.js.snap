// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-breadcrumbs"
>
  <span
    className="jsonschema-inspector-breadcrumbs-icon"
  />
  <span
    className="jsonschema-inspector-breadcrumbs-item has-nested-items"
    key="0"
    onDoubleClick={[Function]}
  >
    Schema One
  </span>
</div>
`;
